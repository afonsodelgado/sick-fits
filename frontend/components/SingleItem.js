import React, { Component } from 'react'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largerImage
    }
  }
`

class SingleItem extends Component {
  render() {
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
        {({ data, loading, error }) => {
          if (loading) return <p>Loading...</p>
          if (error) return <p>Error!</p>

          return (
            <div>
              <p>SingleItem comp</p>
            </div>
          )
        }}
      </Query>
    )
  }
}

export default SingleItem;
