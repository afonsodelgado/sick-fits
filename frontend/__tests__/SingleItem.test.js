import { mount } from 'enzyme'
import { MockedProvider } from 'react-apollo/test-utils'
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem'
import 'jest-styled-components'

const mocks = [
  {
    request: { query: SINGLE_ITEM_QUERY, variables: { id: 'id123' } },
    result: {
      data: {
        item: {
          __typename: 'Item',
          id: 'id123',
          title: 'Fake Item',
          description: 'This is a fake item',
          largerImage: 'item.jpg',
        }
      }
    }
  }
]

describe('<Item />', () => {
  it ('renders and shows loading while fetching data', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="id123" />
      </MockedProvider>
    )
    expect(wrapper.text()).toContain('Loading...')
  })

  it('renders and match snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="id123" />
      </MockedProvider>
    )
    await new Promise(resolve => setTimeout(resolve))
    wrapper.update()
    expect(wrapper.find('SingleItem')).toMatchSnapshot()
  })

  it('renders an error when there is no item', async () => {
    const emptyItemMocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: 'id123' } },
        result: {
          errors: [{ message: 'No item found' }]
        }
      }
    ]

    const wrapper = mount(
      <MockedProvider mocks={emptyItemMocks}>
        <SingleItem id="id123" />
      </MockedProvider>
    )
    await new Promise(resolve => setTimeout(resolve))
    wrapper.update()
    expect(wrapper.find('[data-test="graphql-error"]').text()).toContain('No item found')
    expect(wrapper.find('[data-test="graphql-error"]')).toMatchSnapshot()
  })
});