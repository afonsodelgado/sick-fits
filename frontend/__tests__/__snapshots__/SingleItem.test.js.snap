// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> renders an error when there is no item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  No item found
</p>
`;

exports[`<Item /> renders and match snapshot 1`] = `
.c0 {
  max-width: 1200px;
  margin: 2rem auto;
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
}

.c0 img {
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c0 .details {
  margin: 3rem;
  font-size: 2rem;
}

<SingleItem
  id="id123"
>
  <Query
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "SINGLE_ITEM_QUERY",
            },
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "where",
                      },
                      "value": Object {
                        "fields": Array [
                          Object {
                            "kind": "ObjectField",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "value": Object {
                              "kind": "Variable",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                            },
                          },
                        ],
                        "kind": "ObjectValue",
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "item",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "title",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "description",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "largerImage",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "ID",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 137,
          "start": 0,
        },
      }
    }
    variables={
      Object {
        "id": "id123",
      }
    }
  >
    <div>
      <SingleItem__SingleItemStyles>
        <div
          className="c0"
        >
          <SideEffect(Head)>
            <Head />
          </SideEffect(Head)>
          <img
            alt="Fake Item"
            src="item.jpg"
          />
          <div
            className="details"
          >
            <h2>
              Viewing 
              Fake Item
            </h2>
            <p>
              This is a fake item
            </p>
          </div>
        </div>
      </SingleItem__SingleItemStyles>
    </div>
  </Query>
</SingleItem>
`;
